### PR TITLE
feat: port CircleCI jobs to GH Actions and improve sharness reporting

### LINKED ISSUE #8991: Improve sharness tests failure reports
It's very unclear which test fail.
The error message is not in the failing test (which is just a `grep failed`).
You need to deep dive through logs, CTRL + F `failed` (which will show other unrelated failures often).

### PR DESCRIPTION
**Pull request by galargh**
Resolves #8991
Prepares for resolution of #8804 and #9084

In this PR I port CircleCI jobs to GitHub Actions. The newly added GitHub Actions are marked as `ci/gh-experiment` for now. We want to allow a transition period where both CircleCI and GitHub Actions jobs are running. After the transition period we're going to evaluate correctness (were the results the same?) and performance (are runtimes comparable?).

2 out of the new workflows are running on self-hosted runners: `gobuild` and `sharness`. The self-hosted runners were set up using: https://github.com/pl-strflt/tf-aws-gh-runner. The `kubo` runner configuration can be found here: https://github.com/pl-strflt/tf-aws-gh-runner/blob/013c56d4bf827c60186a38235a126b5418314461/runners.tf#L47-L71

The transition period for the new workflows running on self-hosted runners should be ~1 month. The transition period for the rest of the workflows can be shorter.

## Feature highlight: easy to know why sharness  failed

As part of the migration work, I also improved sharness reporting (to make it usable in GitHub Actions). Sharness runs from GitHub Actions will output HTML reports ([example](https://github.com/ipfs/kubo/actions/runs/3288358400/attempts/1#summary-9003264577)) where both unknown breakages (a.k.a failures) and known breakages (a.k.a errors) are clearly marked, and stdout/stderr can be accesses per testcase/testsuite.

> Screenshot 2022-12-02 at 14-58-51 feat port CircleCI jobs to GH Actions and improve sharness reporting · ipfs_kubo@79f8457

> Screenshot 2022-12-02 at 14-58-07 Unit Test Results

###### Testing
- [x] CI on this PR passes for GitHub Actions
- [x] CI on this PR passes for CircleCI

### CODE REVIEW

File: .circleci/main.yml
Comment (galargh) [inline .circleci/main.yml:147]: This is not required. I only updated Ubuntu version as I was trying to investigate runtime differences between CircleCI and GitHub Actions and wanted to make the environment more similar. It proved to have no affect on runtime.

File: .circleci/main.yml
Comment (galargh) [inline .circleci/main.yml:189]: We were not running expensive tests before. The way the make file is written - https://github.com/ipfs/kubo/blob/master/test/sharness/Rules.mk#L46 - it seems the intention was to execute tests with `TEST_EXPENSIVE=1` though.

File: .github/workflows/sharness.yml
Comment (galargh) [inline .github/workflows/sharness.yml:13]: TODO for @galargh ahead of the merge. Not changing it yet because there is no `ipfs/kubo/.github/workflows/runner.yml` in `master` yet so the CI in this PR would break.

File: test/sharness/README.md
Comment (galargh) [inline test/sharness/README.md:1]: We were using https://github.com/mlafeldt/sharness/ with a git.patch before. I think a repo is better for visibility. I'll try to contribute `feat/junit` back to https://github.com/chriscool/sharness.

File: test/sharness/lib/test-generate-junit-html.sh
Comment (galargh) [inline test/sharness/lib/test-generate-junit-html.sh:6]: We use https://saxon.sourceforge.net/#F11HE to turn JUnit XML output by sharness into HTMLs.

The XML-HTML transforms are defined in https://github.com/pl-strflt/ant/tree/feat/testcase-out/src/etc. We use a fork because the original didn't support `System.out` per testcase (among other things). I'll try to contribute `feat/testcase-out` back to https://github.com/apache/ant.

File: test/sharness/t0110-gateway.sh
Comment (galargh) [inline test/sharness/t0110-gateway.sh:124]: The tests here seem to have been fixed prior to me starting this work.

File: test/sharness/t0125-twonode.sh
Comment (galargh) [inline test/sharness/t0125-twonode.sh:91]: Without the sleep the testcase was being flaky for me. I tried to investigate it further by adding some prints here and there but that always deflaked the test so it got me nowhere. In all honesty, I didn't spend too much time on this.

File: test/sharness/t0160-resolve.sh
Comment (galargh) [inline test/sharness/t0160-resolve.sh:133]: It seems to have been resolved prior to me starting this work.

File: test/sharness/t0300-docker-image.sh
Comment (galargh) [inline test/sharness/t0300-docker-image.sh:84]: Two issues were addresses to fix this case:
- replaced `ipfs add | cut -d' ' -f2` which broke quoting (unescaped `'`) with `ipfs add -q`
- added `tr -d "[:cntrl:]"` to all outs we care about because `docker_exec` returns results `\r`

This go me unblocked which is all I was after for now.

File: .github/workflows/sharness.yml
Comment (galargh) [inline .github/workflows/sharness.yml:30]:
```suggestion
        run: sudo apt install socat net-tools fish libxml2-utils
```

File: .circleci/main.yml
Comment (lidel) [reply .circleci/main.yml:188]: Ack,  tests flagged as expensive live in `test/sharness/t0040-add-and-cat.sh`:

```console
$ rg EXPENSIVE test/sharness
test/sharness/t0040-add-and-cat.sh
588:  test_expect_success EXPENSIVE "generate 100MB file using go-random" '
592:  test_expect_success EXPENSIVE "sha1 of the file looks ok" '
598:  test_expect_success EXPENSIVE "ipfs add $ADD_FLAGS bigfile succeeds" '
602:  test_expect_success EXPENSIVE "ipfs add bigfile output looks good" '
607:  test_expect_success EXPENSIVE "ipfs cat succeeds" '
611:  test_expect_success EXPENSIVE "ipfs cat output looks good" '
616:  test_expect_success EXPENSIVE "ipfs cat output hashed looks good" '
621:  test_expect_success FUSE,EXPENSIVE "cat ipfs/bigfile succeeds" '
625:  test_expect_success FUSE,EXPENSIVE "cat ipfs/bigfile looks good" '

test/sharness/lib/test-lib.sh
95:test "$TEST_EXPENSIVE" = 1 && test_set_prereq EXPENSIVE
111:  echo '# TEST_EXPENSIVE='"$TEST_EXPENSIVE"

test/sharness/Rules.mk
69:test_sharness_expensive: export TEST_EXPENSIVE=1
```

They seem to add ~5 seconds to the runtime on my PC, I think that is ok. 
But maybe remove the makefile magic and have explicit  `TEST_EXPENSIVE=1` in the workflow, like we do `TEST_NO_PLUGIN`?

File: .github/workflows/runner.yml
Comment (lidel) [inline .github/workflows/runner.yml:24]: :thought_balloon: 
@galargh @BigLep 
After we merge this PR, how long we need to run old setup on circleci and github actions with self-hosted worker on AWS in parallel to be able to compare costs? 1-2 months?

File: test/sharness/lib/test-generate-junit-html.sh
Comment (lidel) [reply test/sharness/lib/test-generate-junit-html.sh:6]: nit: can we add `actions/cache` based on the hash of `test/sharness/lib/test-generate-junit-html.sh`, so download of these dependencies happens only once?

**Review by lidel — CHANGES_REQUESTED** [review]
@galargh thank you for investing time into this. 

I'd like us to merge this, and run both for a while – ideally a full release cycle – to compare speed, developer experience, and cost. 

Small asks/suggestions below, but other than that lgm. let me know when this is ready for merge (resolve your own TODOs below), and I'll make final look and merge.

File: .github/workflows/runner.yml
Comment (BigLep) [reply .github/workflows/runner.yml:24]: If I can get a ballpark estimate on the cost that will help me determine whether this is even worth paying attention.  Making our setup easier to support/maintain and let external developers more easily fork and run CI is worth a lot.

Comment (galargh) [reply .github/workflows/runner.yml:24]: Really rough estimate here and I totally might have missed something but I think it's gonna work out to something like this:
- GitHub Actions: ~$99 / month (30 days * $2 for ALB, Gateway, Lambda, etc. + [261 runs in the past 30 days](https://app.circleci.com/insights/github/ipfs/kubo/workflows/test/overview?branch=b1d8acd4-8701-4c76-8a87-7ff2320aa95f) * $0.15 for EC2, EBS, S3, etc.)
- CircleCI: ~$468 / month ([780427 credits used in the past 30 days](https://app.circleci.com/insights/github/ipfs/kubo/workflows/test/overview?branch=b1d8acd4-8701-4c76-8a87-7ff2320aa95f) * [$0.0006 per credit](https://discuss.circleci.com/t/performance-plan-billing-update/39502))

File: .circleci/main.yml
Comment (galargh) [reply .circleci/main.yml]: Done! I replaced `test_sharness_expensive` and `test_sharness_short` directives with a single `test_sharness` one. Now expensive sharness tests are controlled only via `TEST_EXPENSIVE` variable.

File: .github/workflows/sharness.yml
Comment (galargh) [reply .github/workflows/sharness.yml:13]: Actually, I'll change it in a subsequent PR. That way this one will be green when we merge it and `master` will be green after the merge.

File: .github/workflows/gobuild.yml
Comment (galargh) [reply .github/workflows/gobuild.yml:13]: Actually, I'll change it in a subsequent PR. That way this one will be green when we merge it and `master` will be green after the merge.

File: test/sharness/lib/test-generate-junit-html.sh
Comment (galargh) [reply test/sharness/lib/test-generate-junit-html.sh:6]: Good idea! Done :)

File: .github/workflows/build.yml
Comment (galargh) [reply .github/workflows/build.yml]: Done :)

File: .github/workflows/build.yml
Comment (galargh) [reply .github/workflows/build.yml]: Done :) It'd be nice to have a single source for Go version overall but that's out of scope for this PR.

**Comment by galargh** [comment]
@lidel Thanks for the review :) I think I incorporated all of the feedback now. I'll deal with updating actions references in a subsequent PR (eg. https://github.com/ipfs/kubo/pull/9355#discussion_r1039554137).

**Comment by lidel** [comment]
I've fixed some flaky tests and bumped this PR to have changes from master.
 Sadly run out of time before long weekend, but will review next week.

File: .github/workflows/sharness.yml
Comment (galargh) [inline .github/workflows/sharness.yml:80]: ❤️

**Review by lidel — APPROVED** [review]
Thank you so much @galargh, 


Pushed small tweaks and have one concern below, but nothing blocking this from being merged.

Let's see how this behaves during 0.18 and 0.19 releases, 
and decide after 0.19 if we are ok with removal of CircleCI.

**Pull request merged**
